docs/7-algorithms-and-data-structures-part-1/hometask/Dijkstra.ts

```diff
@@ -0,0 +1,98 @@
+
+interface DijkstraI {
+    findShortestPath(vertex1: string, vertex2: string): string[];
+    findPointsOfShortestWay(start: string, finish: string, weight: number): string[];
+}
+
+class NodeVertex {
+    nameOfVertex: string;
+    weight: number;
+}
+
+class Vertex {
+    name: string;
+    nodes: NodeVertex[];
+    weight: number;
+
+    constructor(theName: string, theNodes: NodeVertex[], theWeight: number) {
+        this.name = theName;
+        this.nodes = theNodes;
+        this.weight = theWeight;
+    }
+}
+
+export class Dijkstra implements DijkstraI {
+
+    vertices: any;
+    constructor() {
+        this.vertices = {};
+    }
+
+    addVertex(vertex: Vertex): void {
+        this.vertices[vertex.name] = vertex;
+    }
+
+    findPointsOfShortestWay(start: string, finish: string, weight: number): string[] {
+
+        let nextVertex: string = finish;
+        let arrayWithVertex: string[] = [];
+        while (nextVertex !== start) {
+
+            let minWeigth: number = Number.MAX_VALUE;
+            let minVertex: string = "";
+            for (let i of this.vertices[nextVertex].nodes) {
+                if (i.weight + this.vertices[i.nameOfVertex].weight < minWeigth) {
+                    minWeigth = this.vertices[i.nameOfVertex].weight;
+                    minVertex = i.nameOfVertex;
+                }
+            }
+            arrayWithVertex.push(minVertex);
+            nextVertex = minVertex;
+        }
+        return arrayWithVertex;
+    }
+
+
+    findShortestPath(start: string, finish: string): string[] {
+
+        let nodes: any = {};
+        let visitedVertex: string[] = [];
+
+        for (let i in this.vertices) {
+            if (this.vertices[i].name === start) {
+                this.vertices[i].weight = 0;
+
+            } else {
+                this.vertices[i].weight = Number.MAX_VALUE;
+            }
+            nodes[this.vertices[i].name] = this.vertices[i].weight;
+        }
+
+        while (Object.keys(nodes).length !== 0) {
+            let sortedVisitedByWeight: string[] = Object.keys(nodes).sort((a, b) => this.vertices[a].weight - this.vertices[b].weight);
+            let currentVertex: Vertex = this.vertices[sortedVisitedByWeight[0]];
+            for (let j of currentVertex.nodes) {
+                const calculateWeight: number = currentVertex.weight + j.weight;
+                if (calculateWeight < this.vertices[j.nameOfVertex].weight) {
+                    this.vertices[j.nameOfVertex].weight = calculateWeight;
+                }
+            }
+            delete nodes[sortedVisitedByWeight[0]];
+        }
+        const finishWeight: number = this.vertices[finish].weight;
+        let arrayWithVertex: string[] = this.findPointsOfShortestWay(start, finish, finishWeight).reverse();
+        arrayWithVertex.push(finish, finishWeight.toString());
+        return arrayWithVertex;

```

Comment:
Overall it's okay, but too much of unnecessary computations.

Here is another approach:

```ts
function dijkstra(graph: WeightedGraph, startVertex: Vertex) {
    const paths = new Map<Vertex, number>(); // Map of Vertex -> number(distance to the vertex from startVertex)
    const visited = new Set<Vertex>();

    let current = startVertex;
    paths.set(startVertex, 0); // distance to the startVertex is 0

    while (visited.size < graph.vertices.length) {
        let smallest: { vertex: Vertex, distance: number } | null = null;
        visited.add(current);

        current.edges.forEach((e: Edge) => {
            // if we been in the target node already: just to skip 
            if (visited.has(e.to)) return;

            const distance = paths.get(current) + e.weight;
            if (distance < (paths.get(e.to) ?? Infinity)) {
                paths.set(e.to, distance);
            }
            if (smallest == null || smallest.distance > distance) {
                smallest = {
                    vertex: e.to,
                    distance,
                };
            }
        });
        // indicates that there is no any paths from this node, we are at the finish
        if (smallest == null) break;
        current = smallest.vertex;
    }
    return paths;
}
```

Classes and interfaces here a little bit different, but I think you will get an idea